syntax = "proto3";

package caffa.rpc;

option java_multiple_files = true;
option java_package = "org.caffa.rpc";
option optimize_for = SPEED;

import "ObjectService.proto";

service FieldAccess
{
    rpc SetValue( stream SetterChunk ) returns( SetterReply ) {}
    rpc GetValue( FieldRequest ) returns( stream GetterReply ) {}
}

message IntArray
{
    repeated int32 data = 1 [packed=true];
}

message Int64Array
{
    repeated int64 data = 1 [packed=true];
}

message UInt64Array
{
    repeated uint64 data = 1 [packed=true];
}

message DoubleArray
{
    repeated double data = 1  [packed=true];
}

message FloatArray
{
    repeated float data = 1 [packed=true];
}

message StringArray
{
    repeated string data = 1;
}

message FieldRequest
{
    Object self   = 1;
    string method = 2;
    Object params = 3;
}

message SetterRequest
{
    FieldRequest request     = 1;
    int32        value_count = 2;
}

message SetterChunk
{
    oneof data
    {
        SetterRequest set_request  = 1;
        string        scalar       = 2;
        IntArray      ints         = 3;
        DoubleArray   doubles      = 4;
        StringArray   strings      = 5;
        FloatArray    floats       = 6;
        Int64Array    int64s       = 7;
        UInt64Array   uint64s      = 8;
    }
}

message GetterReply
{
    oneof data
    {
        string      scalar = 1;
        IntArray    ints    = 2;
        DoubleArray doubles = 3;
        StringArray strings = 4;
        FloatArray  floats  = 5;
        Int64Array  int64s  = 6;
        UInt64Array uint64s = 7;

    }
}

message SetterReply
{
    int64 value_count = 1;
}
